fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVTask.java

```diff
@@ -272,6 +285,220 @@ public void run() throws JobException {
+                            executor.getParsedStmt(), executor.getQueryStatisticsForAuditLog(),
+                            true);
+                }
+            }
+        }
+
+        protected abstract Command getDataRefreshCommand(
+                Set<String> refreshPartitionNames, Map<TableIf, String> tableWithPartKey) throws UserException;
+
+        public List<String> calculateNeedRefreshPartitions(MTMVRefreshContext context)
+                throws AnalysisException {
+            List<String> needRefreshPartitionsByManual = calculateNeedRefreshPartitionsByManual(context);
+            if (!needRefreshPartitionsByManual.isEmpty()) {
+                return needRefreshPartitionsByManual;
+            }
+            return calculateNeedRefreshPartitionsInternal(context);
+        }
+
+        protected List<String> calculateNeedRefreshPartitionsInternal(MTMVRefreshContext context)
+                throws AnalysisException {
+            // if refreshMethod is COMPLETE, we must FULL refresh, avoid external table MTMV always not refresh
+            if (mtmv.getRefreshInfo().getRefreshMethod() == RefreshMethod.COMPLETE) {
+                return Lists.newArrayList(mtmv.getPartitionNames());
+            }
+            // check if data is fresh
+            // We need to use a newly generated relationship and cannot retrieve it using mtmv.getRelation()
+            // to avoid rebuilding the baseTable and causing a change in the tableId
+            boolean fresh = MTMVPartitionUtil.isMTMVSync(context, relation.getBaseTablesOneLevel(),
+                    mtmv.getExcludedTriggerTables());
+            if (fresh) {
+                return Lists.newArrayList();
+            }
+            // current, if partitionType is SELF_MANAGE, we can only FULL refresh
+            if (mtmv.getMvPartitionInfo().getPartitionType() == MTMVPartitionType.SELF_MANAGE) {
+                return Lists.newArrayList(mtmv.getPartitionNames());
+            }
+            // We need to use a newly generated relationship and cannot retrieve it using mtmv.getRelation()
+            // to avoid rebuilding the baseTable and causing a change in the tableId
+            return MTMVPartitionUtil.getMTMVNeedRefreshPartitions(context, relation.getBaseTablesOneLevel());
+        }
+
+        public List<String> calculateNeedRefreshPartitionsByManual(MTMVRefreshContext context) {
+            // check whether the user manually triggers it
+            if (taskContext.getTriggerMode() == MTMVTaskTriggerMode.MANUAL) {
+                if (taskContext.isComplete()) {
+                    return Lists.newArrayList(mtmv.getPartitionNames());
+                } else if (!CollectionUtils.isEmpty(taskContext.getPartitions())) {
+                    return taskContext.getPartitions();
+                }
+            }
+            return Lists.newArrayList();
+        }
+    }
+
+    private class MTMVFullDataRefreshExec extends MTMVDataRefreshExec {
```

Comment:
I think the name `MTMVFullDataRefreshExec` is not suitable. Because "partition level refresh" is also a kind of `incremental` refresh. How about `PartitionRefresh` and `SnapshotRefresh`?